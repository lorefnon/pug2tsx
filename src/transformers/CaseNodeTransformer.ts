import * as Pug from "../pug";
import * as t from "@babel/types";
import { Transformer } from "./Transformer";
import { compact, flatten } from "lodash";
import { parseExpression } from "@babel/parser";
import { ErrorCode } from "../CompilationError";
import { FunctionBodyTransformer } from "./FunctionBodyTransformer";
import { PugNodeTransformer } from "./PugNodeTransformer";

export class CaseNodeTransformer extends Transformer<Pug.CaseNode, t.Node> {
    transform(): void {
        try {
            this.output = t.callExpression(
                t.arrowFunctionExpression(
                    [],
                    t.blockStatement([
                        t.switchStatement(
                            parseExpression(this.input.expr),
                            compact(
                                this.input.block.nodes.map(n => {
                                    if (!Pug.isType<Pug.WhenNode>(n, Pug.Type.When)) {
                                        this.pushError({
                                            reasons: ["Expected only when node in case statement body"],
                                            isFatal: true,
                                            code: ErrorCode.UnsupportedSyntaxError,
                                        });
                                        return;
                                    }
                                    const transformed = flatten(
                                        compact(n.block.nodes.map(n => this.delegateTo(PugNodeTransformer, n))),
                                    );
                                    const body = this.delegateTo(FunctionBodyTransformer, transformed);
                                    if (!body) return;
                                    if (n.expr === "default") return t.switchCase(null, body);
                                    else return t.switchCase(parseExpression(n.expr), body);
                                }),
                            ),
                        ),
                    ]),
                ),
                [],
            );
            this.commentLineNumber(this.output, this.input)
        } catch (e) {
            this.pushError({
                code: ErrorCode.UnsupportedSyntaxError,
                originalError: e,
                isFatal: true,
            });
        }
    }
}
