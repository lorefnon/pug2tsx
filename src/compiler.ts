import { FileTransformationContext } from "./transformers/FileTransformationContext";
import { FileTransfomer } from "./transformers/FileTransformer";

export interface CompilationOptions {
    filePath?: string;
    skipLineAnnotations?: boolean
}

export const transpile = (inputContent: string, options: CompilationOptions = {}) => {
    const context = new FileTransformationContext(
        options.filePath,
        options.skipLineAnnotations
    );
    const transformer = new FileTransfomer(inputContent, context);
    transformer.transform();
    return {
        result: transformer.output,
        errors: context.errors,
    };
};
