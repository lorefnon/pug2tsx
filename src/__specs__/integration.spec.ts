import { transpile } from "../compiler";
import dedent from "dedent";
import { isEmpty } from "lodash";
import { parsePug } from "../transformers/FileTransformer";

test("Pug inside arrow functions", () => {
    const template = dedent`
    const HelloWorld = () =>
        #bar.foo
            .foo__node.foo__node--aspect(data-bar="baz")
                | Hello world`;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR = transpile(template, {
        defaultExportName: "SampleComponent",
    });
    expect(transpiledR).toMatchSnapshot();
    expect(isEmpty(transpiledR.errors)).toBe(true);
});

test("Pug inside old-school functions", () => {
    const template = dedent`
    function HelloWorld(props: SomeProps)
        #bar.foo
            .foo__node.foo__node--aspect(data-bar="baz")
                | Hello world`;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR = transpile(template, {
        defaultExportName: "SampleComponent",
    });
    expect(transpiledR).toMatchSnapshot();
    expect(isEmpty(transpiledR.errors)).toBe(true);
});

test("Interpolated Code", () => {
    const template = dedent`
    const Hello = () =>
        - const foo = "bar";
        #bar.foo
            .foo__node.foo__node--aspect(data-bar="baz")= foo
        div.hello
            = \`Hello \${foo}\``;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR = transpile(template, {
        defaultExportName: "SampleComponent",
    });
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});

test("Interleaved unbuffered code", () => {
    const template = dedent`
    const Hello = () =>
        - const foo = "bar";
        #bar.foo
            .foo__node.foo__node--aspect Test
        - const bar = "baz";
        div.hello
            = \`Hello \${foo} \${bar}\``;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR = transpile(template, {
        defaultExportName: "SampleComponent",
    });
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});

test("Unbuffered multiline code", () => {
    const template = dedent`
    -
        const obj = {
            a: 10,
            b: 20
        }

    const Hello = () =>
        #bar.foo = obj.a`;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR = transpile(template, {
        defaultExportName: "SampleComponent",
    });
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});

test("Sanitization of buffered code", () => {
    const template = dedent`
    const Hello = () =>
        #bar.foo
            = '<div>Hello</div>'`;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR = transpile(template, {
        defaultExportName: "SampleComponent",
    });
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});

test("Assigning elements", () => {
    const template = dedent`
    const Hello = () =>
        - const foo =
          div.hello
            | World

        div.hello
            SomeTag(name=foo)`;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR: any = transpile(template, {
        defaultExportName: "TestComponent",
    });
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});


test("Render props", () => {
    const template = dedent`
    const Hello = () =>
        AppContext.Consumer
            = (app) =>
                if app.repositoryRoot
                    #workspace Success
                else
                    #workspace Failure`;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR: any = transpile(template);
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});

test("Multiple children in nested block", () => {
    const template = dedent`
        - const Foo = () =>
            #foo.bar
                | Foo Bar
            #bar.baz
                | Bar baz
        - const Bar = () =>
            #foo.bar.baz
                | Foo Bar
            #bar.baz
                | Bar baz
        - export default Foo
    `;
    expect(parsePug(template)).toMatchSnapshot();
    const transpiledR: any = transpile(template);
    expect(isEmpty(transpiledR.errors)).toBe(true);
    expect(transpiledR).toMatchSnapshot();
});
