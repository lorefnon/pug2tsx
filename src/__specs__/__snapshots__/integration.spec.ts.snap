// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assigning elements 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "column": 7,
            "line": 1,
            "type": "Text",
            "val": "Hello = () =>",
          },
          Object {
            "block": Object {
              "line": 3,
              "nodes": Array [
                Object {
                  "attributeBlocks": Array [],
                  "attrs": Array [
                    Object {
                      "column": 10,
                      "line": 3,
                      "mustEscape": false,
                      "name": "class",
                      "val": "'hello'",
                    },
                  ],
                  "block": Object {
                    "line": 3,
                    "nodes": Array [
                      Object {
                        "column": 11,
                        "line": 4,
                        "type": "Text",
                        "val": "World",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 7,
                  "isInline": false,
                  "line": 3,
                  "name": "div",
                  "selfClosing": false,
                  "type": "Tag",
                },
              ],
              "type": "Block",
            },
            "buffer": false,
            "column": 5,
            "isInline": false,
            "line": 2,
            "mustEscape": false,
            "type": "Code",
            "val": "const foo =",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 8,
                "line": 6,
                "mustEscape": false,
                "name": "class",
                "val": "'hello'",
              },
            ],
            "block": Object {
              "line": 6,
              "nodes": Array [
                Object {
                  "attributeBlocks": Array [],
                  "attrs": Array [
                    Object {
                      "column": 17,
                      "line": 7,
                      "mustEscape": true,
                      "name": "name",
                      "val": "foo",
                    },
                  ],
                  "block": Object {
                    "line": 7,
                    "nodes": Array [],
                    "type": "Block",
                  },
                  "column": 9,
                  "isInline": false,
                  "line": 7,
                  "name": "SomeTag",
                  "selfClosing": false,
                  "type": "Tag",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 6,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "const",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Assigning elements 2`] = `
Object {
  "errors": Array [],
  "result": "/*@pug:L1*/
<const>
  Hello = () =>
  {(() => {
    /*@pug:L2*/
    const foo = (
      /*@pug:L3*/
      <div className={\\"hello\\"}>World</div>
    );
    return (
      /*@pug:L6*/
      <div className={\\"hello\\"}>
        /*@pug:L7*/
        <SomeTag name={foo}></SomeTag>
      </div>
    );
  })()}
</const>;
",
}
`;

exports[`Interleaved unbuffered code 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "column": 7,
            "line": 1,
            "type": "Text",
            "val": "Hello = () =>",
          },
          Object {
            "buffer": false,
            "column": 5,
            "isInline": false,
            "line": 2,
            "mustEscape": false,
            "type": "Code",
            "val": "const foo = \\"bar\\";",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 3,
                "mustEscape": false,
                "name": "id",
                "val": "'bar'",
              },
              Object {
                "column": 9,
                "line": 3,
                "mustEscape": false,
                "name": "class",
                "val": "'foo'",
              },
            ],
            "block": Object {
              "line": 3,
              "nodes": Array [
                Object {
                  "attributeBlocks": Array [],
                  "attrs": Array [
                    Object {
                      "column": 9,
                      "line": 4,
                      "mustEscape": false,
                      "name": "class",
                      "val": "'foo__node'",
                    },
                    Object {
                      "column": 19,
                      "line": 4,
                      "mustEscape": false,
                      "name": "class",
                      "val": "'foo__node--aspect'",
                    },
                  ],
                  "block": Object {
                    "line": 4,
                    "nodes": Array [
                      Object {
                        "column": 38,
                        "line": 4,
                        "type": "Text",
                        "val": "Test",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "isInline": false,
                  "line": 4,
                  "name": "div",
                  "selfClosing": false,
                  "type": "Tag",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 3,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
          Object {
            "buffer": false,
            "column": 5,
            "isInline": false,
            "line": 5,
            "mustEscape": false,
            "type": "Code",
            "val": "const bar = \\"baz\\";",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 8,
                "line": 6,
                "mustEscape": false,
                "name": "class",
                "val": "'hello'",
              },
            ],
            "block": Object {
              "line": 6,
              "nodes": Array [
                Object {
                  "buffer": true,
                  "column": 9,
                  "isInline": false,
                  "line": 7,
                  "mustEscape": true,
                  "type": "Code",
                  "val": "\`Hello \\\\\${foo} \\\\\${bar}\`",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 6,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "const",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Interleaved unbuffered code 2`] = `
Object {
  "errors": Array [],
  "result": "/*@pug:L1*/
<const>
  Hello = () =>
  {(() => {
    /*@pug:L2*/
    const foo = \\"bar\\";
    return (
      <>
        /*@pug:L3*/
        <div id={\\"bar\\"} className={\\"foo\\"}>
          /*@pug:L4*/
          <div className={[\\"foo__node\\", \\"foo__node--aspect\\"].join(\\" \\")}>
            Test
          </div>
        </div>
        {(() => {
          /*@pug:L5*/
          const bar = \\"baz\\";
          return (
            /*@pug:L6*/
            <div className={\\"hello\\"}>{
              /*@pug:L7*/
              \`Hello \\\\\${foo} \\\\\${bar}\`
            }</div>
          );
        })()}
      </>
    );
  })()}
</const>;
",
}
`;

exports[`Interpolated Code 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "column": 7,
            "line": 1,
            "type": "Text",
            "val": "Hello = () =>",
          },
          Object {
            "buffer": false,
            "column": 5,
            "isInline": false,
            "line": 2,
            "mustEscape": false,
            "type": "Code",
            "val": "const foo = \\"bar\\";",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 3,
                "mustEscape": false,
                "name": "id",
                "val": "'bar'",
              },
              Object {
                "column": 9,
                "line": 3,
                "mustEscape": false,
                "name": "class",
                "val": "'foo'",
              },
            ],
            "block": Object {
              "line": 3,
              "nodes": Array [
                Object {
                  "attributeBlocks": Array [],
                  "attrs": Array [
                    Object {
                      "column": 9,
                      "line": 4,
                      "mustEscape": false,
                      "name": "class",
                      "val": "'foo__node'",
                    },
                    Object {
                      "column": 19,
                      "line": 4,
                      "mustEscape": false,
                      "name": "class",
                      "val": "'foo__node--aspect'",
                    },
                    Object {
                      "column": 38,
                      "line": 4,
                      "mustEscape": true,
                      "name": "data-bar",
                      "val": "\\"baz\\"",
                    },
                  ],
                  "block": Object {
                    "line": 4,
                    "nodes": Array [
                      Object {
                        "buffer": true,
                        "column": 53,
                        "isInline": true,
                        "line": 4,
                        "mustEscape": true,
                        "type": "Code",
                        "val": "foo",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "isInline": false,
                  "line": 4,
                  "name": "div",
                  "selfClosing": false,
                  "type": "Tag",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 3,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 8,
                "line": 5,
                "mustEscape": false,
                "name": "class",
                "val": "'hello'",
              },
            ],
            "block": Object {
              "line": 5,
              "nodes": Array [
                Object {
                  "buffer": true,
                  "column": 9,
                  "isInline": false,
                  "line": 6,
                  "mustEscape": true,
                  "type": "Code",
                  "val": "\`Hello \\\\\${foo}\`",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 5,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "const",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Interpolated Code 2`] = `
Object {
  "errors": Array [],
  "result": "/*@pug:L1*/
<const>
  Hello = () =>
  {(() => {
    /*@pug:L2*/
    const foo = \\"bar\\";
    return (
      <>
        /*@pug:L3*/
        <div id={\\"bar\\"} className={\\"foo\\"}>
          /*@pug:L4*/
          <div
            data-bar={\\"baz\\"}
            className={[\\"foo__node\\", \\"foo__node--aspect\\"].join(\\" \\")}
          >
            {
              /*@pug:L4*/
              foo
            }
          </div>
        </div>
        /*@pug:L5*/
        <div className={\\"hello\\"}>{
          /*@pug:L6*/
          \`Hello \\\\\${foo}\`
        }</div>
      </>
    );
  })()}
</const>;
",
}
`;

exports[`Multiple children in nested block 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "block": Object {
        "line": 2,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 2,
                "mustEscape": false,
                "name": "id",
                "val": "'foo'",
              },
              Object {
                "column": 9,
                "line": 2,
                "mustEscape": false,
                "name": "class",
                "val": "'bar'",
              },
            ],
            "block": Object {
              "line": 2,
              "nodes": Array [
                Object {
                  "column": 11,
                  "line": 3,
                  "type": "Text",
                  "val": "Foo Bar",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 4,
                "mustEscape": false,
                "name": "id",
                "val": "'bar'",
              },
              Object {
                "column": 9,
                "line": 4,
                "mustEscape": false,
                "name": "class",
                "val": "'baz'",
              },
            ],
            "block": Object {
              "line": 4,
              "nodes": Array [
                Object {
                  "column": 11,
                  "line": 5,
                  "type": "Text",
                  "val": "Bar baz",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 4,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const Foo = () =>",
    },
    Object {
      "block": Object {
        "line": 7,
        "nodes": Array [
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 7,
                "mustEscape": false,
                "name": "id",
                "val": "'foo'",
              },
              Object {
                "column": 9,
                "line": 7,
                "mustEscape": false,
                "name": "class",
                "val": "'bar'",
              },
              Object {
                "column": 13,
                "line": 7,
                "mustEscape": false,
                "name": "class",
                "val": "'baz'",
              },
            ],
            "block": Object {
              "line": 7,
              "nodes": Array [
                Object {
                  "column": 11,
                  "line": 8,
                  "type": "Text",
                  "val": "Foo Bar",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 7,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 9,
                "mustEscape": false,
                "name": "id",
                "val": "'bar'",
              },
              Object {
                "column": 9,
                "line": 9,
                "mustEscape": false,
                "name": "class",
                "val": "'baz'",
              },
            ],
            "block": Object {
              "line": 9,
              "nodes": Array [
                Object {
                  "column": 11,
                  "line": 10,
                  "type": "Text",
                  "val": "Bar baz",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 9,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 6,
      "mustEscape": false,
      "type": "Code",
      "val": "const Bar = () =>",
    },
    Object {
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 11,
      "mustEscape": false,
      "type": "Code",
      "val": "export default Foo",
    },
  ],
  "type": "Block",
}
`;

exports[`Multiple children in nested block 2`] = `
Object {
  "errors": Array [],
  "result": "/*@pug:L1*/
const Foo = () => {
  return (
    <>
      /*@pug:L2*/
      <div id={\\"foo\\"} className={\\"bar\\"}>
        Foo Bar
      </div>
      /*@pug:L4*/
      <div id={\\"bar\\"} className={\\"baz\\"}>
        Bar baz
      </div>
    </>
  );
};

/*@pug:L6*/
const Bar = () => {
  return (
    <>
      /*@pug:L7*/
      <div id={\\"foo\\"} className={[\\"bar\\", \\"baz\\"].join(\\" \\")}>
        Foo Bar
      </div>
      /*@pug:L9*/
      <div id={\\"bar\\"} className={\\"baz\\"}>
        Bar baz
      </div>
    </>
  );
};

/*@pug:L11*/
export default Foo;
",
}
`;

exports[`Pug inside arrow functions 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "column": 7,
            "line": 1,
            "type": "Text",
            "val": "HelloWorld = () =>",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 2,
                "mustEscape": false,
                "name": "id",
                "val": "'bar'",
              },
              Object {
                "column": 9,
                "line": 2,
                "mustEscape": false,
                "name": "class",
                "val": "'foo'",
              },
            ],
            "block": Object {
              "line": 2,
              "nodes": Array [
                Object {
                  "attributeBlocks": Array [],
                  "attrs": Array [
                    Object {
                      "column": 9,
                      "line": 3,
                      "mustEscape": false,
                      "name": "class",
                      "val": "'foo__node'",
                    },
                    Object {
                      "column": 19,
                      "line": 3,
                      "mustEscape": false,
                      "name": "class",
                      "val": "'foo__node--aspect'",
                    },
                    Object {
                      "column": 38,
                      "line": 3,
                      "mustEscape": true,
                      "name": "data-bar",
                      "val": "\\"baz\\"",
                    },
                  ],
                  "block": Object {
                    "line": 3,
                    "nodes": Array [
                      Object {
                        "column": 15,
                        "line": 4,
                        "type": "Text",
                        "val": "Hello world",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "isInline": false,
                  "line": 3,
                  "name": "div",
                  "selfClosing": false,
                  "type": "Tag",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "const",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Pug inside arrow functions 2`] = `
Object {
  "errors": Array [],
  "result": "/*@pug:L1*/
<const>
  HelloWorld = () => /*@pug:L2*/
  <div id={\\"bar\\"} className={\\"foo\\"}>
    /*@pug:L3*/
    <div
      data-bar={\\"baz\\"}
      className={[\\"foo__node\\", \\"foo__node--aspect\\"].join(\\" \\")}
    >
      Hello world
    </div>
  </div>
</const>;
",
}
`;

exports[`Pug inside old-school functions 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "column": 10,
            "line": 1,
            "type": "Text",
            "val": "HelloWorld(props: SomeProps)",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 2,
                "mustEscape": false,
                "name": "id",
                "val": "'bar'",
              },
              Object {
                "column": 9,
                "line": 2,
                "mustEscape": false,
                "name": "class",
                "val": "'foo'",
              },
            ],
            "block": Object {
              "line": 2,
              "nodes": Array [
                Object {
                  "attributeBlocks": Array [],
                  "attrs": Array [
                    Object {
                      "column": 9,
                      "line": 3,
                      "mustEscape": false,
                      "name": "class",
                      "val": "'foo__node'",
                    },
                    Object {
                      "column": 19,
                      "line": 3,
                      "mustEscape": false,
                      "name": "class",
                      "val": "'foo__node--aspect'",
                    },
                    Object {
                      "column": 38,
                      "line": 3,
                      "mustEscape": true,
                      "name": "data-bar",
                      "val": "\\"baz\\"",
                    },
                  ],
                  "block": Object {
                    "line": 3,
                    "nodes": Array [
                      Object {
                        "column": 15,
                        "line": 4,
                        "type": "Text",
                        "val": "Hello world",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "isInline": false,
                  "line": 3,
                  "name": "div",
                  "selfClosing": false,
                  "type": "Tag",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "function",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Pug inside old-school functions 2`] = `
Object {
  "errors": Array [],
  "result": "/*@pug:L1*/
<function>
  HelloWorld(props: SomeProps) /*@pug:L2*/
  <div id={\\"bar\\"} className={\\"foo\\"}>
    /*@pug:L3*/
    <div
      data-bar={\\"baz\\"}
      className={[\\"foo__node\\", \\"foo__node--aspect\\"].join(\\" \\")}
    >
      Hello world
    </div>
  </div>
</function>;
",
}
`;

exports[`Render props 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "column": 7,
            "line": 1,
            "type": "Text",
            "val": "Hello = () =>",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 15,
                "line": 2,
                "mustEscape": false,
                "name": "class",
                "val": "'Consumer'",
              },
            ],
            "block": Object {
              "line": 2,
              "nodes": Array [
                Object {
                  "block": Object {
                    "line": 4,
                    "nodes": Array [
                      Object {
                        "alternate": Object {
                          "line": 7,
                          "nodes": Array [
                            Object {
                              "attributeBlocks": Array [],
                              "attrs": Array [
                                Object {
                                  "column": 17,
                                  "line": 7,
                                  "mustEscape": false,
                                  "name": "id",
                                  "val": "'workspace'",
                                },
                              ],
                              "block": Object {
                                "line": 7,
                                "nodes": Array [
                                  Object {
                                    "column": 28,
                                    "line": 7,
                                    "type": "Text",
                                    "val": "Failure",
                                  },
                                ],
                                "type": "Block",
                              },
                              "column": 17,
                              "isInline": false,
                              "line": 7,
                              "name": "div",
                              "selfClosing": false,
                              "type": "Tag",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "consequent": Object {
                          "line": 5,
                          "nodes": Array [
                            Object {
                              "attributeBlocks": Array [],
                              "attrs": Array [
                                Object {
                                  "column": 17,
                                  "line": 5,
                                  "mustEscape": false,
                                  "name": "id",
                                  "val": "'workspace'",
                                },
                              ],
                              "block": Object {
                                "line": 5,
                                "nodes": Array [
                                  Object {
                                    "column": 28,
                                    "line": 5,
                                    "type": "Text",
                                    "val": "Success",
                                  },
                                ],
                                "type": "Block",
                              },
                              "column": 17,
                              "isInline": false,
                              "line": 5,
                              "name": "div",
                              "selfClosing": false,
                              "type": "Tag",
                            },
                          ],
                          "type": "Block",
                        },
                        "line": 4,
                        "test": "app.repositoryRoot",
                        "type": "Conditional",
                      },
                    ],
                    "type": "Block",
                  },
                  "buffer": true,
                  "column": 9,
                  "isInline": false,
                  "line": 3,
                  "mustEscape": true,
                  "type": "Code",
                  "val": "(app) =>",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "AppContext",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "const",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Render props 2`] = `
Object {
  "errors": Array [],
  "result": "/*@pug:L1*/
<const>
  Hello = () => /*@pug:L2*/
  <AppContext className={\\"Consumer\\"}>
    {
      /*@pug:L3*/
      (app) =>
        /*@pug:L4*/
        app.repositoryRoot ? (
          /*@pug:L5*/
          <div id={\\"workspace\\"}>Success</div>
        ) : (
          /*@pug:L7*/
          <div id={\\"workspace\\"}>Failure</div>
        )
    }
  </AppContext>
</const>;
",
}
`;

exports[`Sanitization of buffered code 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [],
      "block": Object {
        "line": 1,
        "nodes": Array [
          Object {
            "column": 7,
            "line": 1,
            "type": "Text",
            "val": "Hello = () =>",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 2,
                "mustEscape": false,
                "name": "id",
                "val": "'bar'",
              },
              Object {
                "column": 9,
                "line": 2,
                "mustEscape": false,
                "name": "class",
                "val": "'foo'",
              },
            ],
            "block": Object {
              "line": 2,
              "nodes": Array [
                Object {
                  "buffer": true,
                  "column": 9,
                  "isInline": false,
                  "line": 3,
                  "mustEscape": true,
                  "type": "Code",
                  "val": "'<div>Hello</div>'",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 1,
      "name": "const",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Sanitization of buffered code 2`] = `
Object {
  "errors": Array [],
  "result": "/*@pug:L1*/
<const>
  Hello = () => /*@pug:L2*/
  <div id={\\"bar\\"} className={\\"foo\\"}>
    {
      /*@pug:L3*/
      \\"<div>Hello</div>\\"
    }
  </div>
</const>;
",
}
`;

exports[`Unbuffered multiline code 1`] = `
Object {
  "line": 0,
  "nodes": Array [
    Object {
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const obj = {
    a: 10,
    b: 20
}
",
    },
    Object {
      "attributeBlocks": Array [],
      "attrs": Array [],
      "block": Object {
        "line": 7,
        "nodes": Array [
          Object {
            "column": 7,
            "line": 7,
            "type": "Text",
            "val": "Hello = () =>",
          },
          Object {
            "attributeBlocks": Array [],
            "attrs": Array [
              Object {
                "column": 5,
                "line": 8,
                "mustEscape": false,
                "name": "id",
                "val": "'bar'",
              },
              Object {
                "column": 9,
                "line": 8,
                "mustEscape": false,
                "name": "class",
                "val": "'foo'",
              },
            ],
            "block": Object {
              "line": 8,
              "nodes": Array [
                Object {
                  "column": 14,
                  "line": 8,
                  "type": "Text",
                  "val": "= obj.a",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 8,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "isInline": false,
      "line": 7,
      "name": "const",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`Unbuffered multiline code 2`] = `
Object {
  "errors": Array [],
  "result": "/*@pug:L1*/
const obj = {
  a: 10,
  b: 20,
};

/*@pug:L7*/
<const>
  Hello = () => /*@pug:L8*/
  <div id={\\"bar\\"} className={\\"foo\\"}>
    = obj.a
  </div>
</const>;
",
}
`;
